LibOS/shim/src/ipc/shim_ipc_nsimpl.h

```diff
@@ -265,37 +270,43 @@ static int __add_range (struct range * r, int off, IDTYPE owner,
     ret = __set_range_bitmap(off, false);
     if (ret == -EEXIST) {
         struct range * tmp;
-        struct hlist_node * pos;
 
-        hlist_for_each_entry(tmp, pos, head, hlist)
+        listp_for_each_entry(tmp, head, hlist)
             if (tmp->offset == off) {
-                hlist_del(&tmp->hlist);
-                list_del(&tmp->list);
-                if (tmp->owner)
+                listp_del(tmp, head, hlist);
+                /* DEP 5/15/17: I am less confident about this, but I 
+                 * believe that, if r->owner is non-NULL, it is owned, and 
+                 * otherwise it is on offered */
+                if (tmp->owner) {
+                    listp_del(tmp, &owned_ranges, list);
```

Comment:
This one may be difficult to judge, but I agree the constraint here is, as long as tmp->owner is NULL, it should be in the owned list.